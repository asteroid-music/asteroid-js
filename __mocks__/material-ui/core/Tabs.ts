export default "Material UI Tabs";