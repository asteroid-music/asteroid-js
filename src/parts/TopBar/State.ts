/**
 * State interface for the <TopBar> component
 * Details for each individual property is given in comments below.
 */
interface TopBarState {
    /** Whether the sidebar should be open */
    sidebarOpen: boolean,
    /** Name of the currently open sidebar */
    sidebarName: string,
    /** Subtabs of the currently open sidebar */
    sidebarSubtabs: string[]
}

export default TopBarState;
