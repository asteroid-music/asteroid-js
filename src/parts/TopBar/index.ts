//Main component
export { default } from './TopBar';

//Props interface
export { default as TopBarProps } from './Props'

