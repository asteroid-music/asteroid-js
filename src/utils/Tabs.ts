/**
 * Interface describing an object with the same elements as SubTabObject
 */
interface SubTabObjectLike {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    component: React.ComponentType<any>;
}

/**
 * Interface describing an object with the same elements as TabObject
 */
interface TabObjectLike {
    /** The name of the tab */
    name: string;

    /** A subtab array containing names of subtabs */
    subtabs: SubTabObjectLike[];
}

class SubTabObject {
    /** The name of the subtab */
    name: string;

    /** The component of the subtab */
    component: React.ComponentType<any>;

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: SubTabObjectLike) {
        this.name = rawJson.name;
        this.component = rawJson.component;
    }

}

/**
 * Class describing named tab, along with strings representing all the subtabs
 * it contains
 */
class TabObject {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    subtabs: SubTabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike) {
        this.name = rawJson.name;
        this.subtabs = rawJson.subtabs.map((item: SubTabObjectLike) => {
            return new SubTabObject(item);
        });
    }

    /**
     * Gets a list of names of subtabs in the tab
     *
     * @returns {string[]} tablist: the name of each subtab
     */
    nameList() {
       return this.subtabs.map((subtab: SubTabObject) => {
            return subtab.name;
       });
    }

    /**
     * Checks if the subtab is present in the tab
     *
     * @param {string} subtab: the name of the subtab to check for
     *
     * @returns {boolean} isin: true if subtab is in the tab; false otherwise
     */
    includes(subtab:string) {
        return this.nameList().includes(subtab);
    }

    /**
     * Returns the subtab specified by the given subtab name,
     *      or 'null' if none exists
     *
     * @param {string} subtab: the name of the subtab to get
     *
     * @returns {SubTabObject | null} subtab: the specified subtab if exists;
     *      'null' otherwise
     */
    get(subtab:string) {
        let objIndex: number = this.nameList().indexOf(subtab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.subtabs[objIndex];
        }
    }
}

/**
 * Class describing an array of named tabs
 */
class TabArray {
    /** Encapsulated array */
    tabs: TabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike[]) {
        this.tabs = rawJson.map((item: TabObjectLike) => {
            return new TabObject(item);
        });
    }

    /**
     * Gets a list of names of tabs in the array
     *
     * @returns {string[]} tablist: the name of each tab
     */
    nameList() {
       return this.tabs.map((tab: TabObject) => {
            return tab.name;
       });
    }

    /**
     * Checks if the tab is present in the array
     *
     * @param {string} tab: the name of the tab to check for
     *
     * @returns {boolean} isin: true if tab is in the tab array;
     *      false otherwise
     */
    includes(tab:string) {
        return this.nameList().includes(tab);
    }

    /**
     * Returns the tab specified by the given tabname, or 'null' if none exists
     *
     * @param {string} tab: the name of the tab to get
     *
     * @returns {TabObject | null} tab: the specified tab if exists;
     *      'null' otherwise
     */
    get(tab:string) {
        let objIndex: number = this.nameList().indexOf(tab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.tabs[objIndex];
        }
    }
}

export { TabObjectLike, TabObject, TabArray }
