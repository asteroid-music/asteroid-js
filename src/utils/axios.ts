import axios from 'axios';

//console.log(asteroidConfig);

declare var __config__: {
    "axios": {
        "baseURL": string,
        "timeout": number
    }
};

let instance = axios.create(__config__.axios || {});

export default instance;
