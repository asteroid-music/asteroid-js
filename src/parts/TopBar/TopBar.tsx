//Import from external module 'react'
import React from 'react';

//Imports from external module '@material-ui/core'
import Tabs from '@material-ui/core/Tabs';
import AppBar from '@material-ui/core/AppBar';
import Tab from '@material-ui/core/Tab';
import Paper from '@material-ui/core/Paper';
import Drawer from '@material-ui/core/Drawer';

//Imports from external module 'react-router'
import { withRouter } from 'react-router-dom';

//Imports from local folder
import TopBarProps from './Props';
import TopBarState from './State';

//Imports from local 'parts'
import SideBar from 'parts/SideBar';

/**
 * <TopBar> React component for navigating Asteroid.
 * Creates a bar at the top of the screen with buttons to access tabs and
 *      subtabs
 */
class TopBar extends React.Component<TopBarProps,TopBarState> {
    constructor(props) {
        super(props);
        this.state = {
            sidebarOpen: false,
            sidebarName: "",
            sidebarSubtabs: []
        };
    }

    render() {
        let location: string | undefined = this.props.location.pathname.split("/")[1];
        return (
            <AppBar position="static">
                {/*Main tab bar*/}
                <Tabs
                    value={location || false}
                    onChange={
                        (event: React.ChangeEvent<{}>, value: string) => {
                            this.setState({
                                sidebarOpen: true,
                                sidebarName: value,
                                sidebarSubtabs: this.props.tabs.get(value)?.nameList() || []
                            });
                        }
                    }
                    variant="scrollable"
                    scrollButtons="on"
                >
                    {this.props.tabs.nameList().map((tabName: string) => {
                        return <Tab
                            key={tabName}
                            value={tabName}
                            label={tabName}
                        />
                    })}
                </Tabs>
                <SideBar
                    open={this.state.sidebarOpen}
                    tabName={this.state.sidebarName}
                    linkNames={this.state.sidebarSubtabs}
                    closeCallback={()=>{this.setState({sidebarOpen: false});}}
                />
            </AppBar>
        );
    }
}

export default withRouter(TopBar);
