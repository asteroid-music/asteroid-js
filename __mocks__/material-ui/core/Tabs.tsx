import React from 'react';
import { serialize as serialise } from 'jest-snapshot/build/utils';
//I know, UK vs US spelling...

const Tabs = props => <div
>
    Material UI Tabs with props:
    {
        serialise(props)
    }
    {props.onChange ? props.children.map((child)=>{
        return React.cloneElement(child,{onChange: (event,value)=>{props.onChange(event,value);}});
    }) : props.children}
</div>

export default Tabs;