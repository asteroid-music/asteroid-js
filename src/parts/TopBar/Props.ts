//Imports from React module
import React from 'react';

//Import from local 'utils'
import { TabArray } from 'utils/Tabs';

/**
 * Interface for props for the <TopBar> React component.
 * Details for each individual property is given in comments below.
 */
interface TopBarProps {
    /** Props for react router */
    match: object,
    history: object,
    location: {hash: string, pathname: string, search: string},
    /** A TabArray object representing the accessible tabs */
    tabs: TabArray;
}

export default TopBarProps
