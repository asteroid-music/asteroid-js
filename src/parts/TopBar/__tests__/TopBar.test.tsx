import React from 'react';
import renderer from 'react-test-renderer';
import TopBar from 'parts/TopBar';
import { TabArray } from 'utils/Tabs';
import { StaticRouter } from 'react-router-dom';
import { serialize as serialise } from 'jest-snapshot/build/utils';
//I know, UK vs US spelling...

/**
 * TopBar.tsx imports:
 *  - React
 *  - Components from MUI (pre-mocked)
 *  - withRouter from react-router-dom
 *  - SideBar
 * 
 * The following should be mocked:
 *  - withRouter
 *  - SideBar
 * 
 * As this component contains the following interactions:
 *  - Clicking a tab button
 * it should test the interaction with each
 */

jest.mock("parts/SideBar",()=>"AsteroidSideBar");

const testTabs = new TabArray([
    {
        name: "TestTab",
        subtabs: [
            {name: "TestSubtab", component: TopBar},
            {name: "TestSubtabTwo", component: TopBar},
            {name: "TestSubtab3", component: TopBar},
            {name: "TestSubtab-4", component: TopBar},
        ]
    }
]);

describe('<TopBar />',()=>{
    //Basic rendering with minimal props
    it("renders correctly",()=>{
        let tree = renderer.create(<StaticRouter>
            <TopBar
                tabs={testTabs}
            />
        </StaticRouter>)
        const json = tree.toJSON();
        expect(json).toMatchSnapshot();
    });

    //Interaction with a tab button
    it("opens the correct sidebar when a tab button is pressed",()=>{
        let tree = renderer.create(<StaticRouter>
            <TopBar
                tabs={testTabs}
            />
        </StaticRouter>)
        //Interact with the button
        tree.root.findByType('button').props.onClick();
        //Match with snapshot
        const json = tree.toJSON();
        expect(json).toMatchSnapshot();
});

});
