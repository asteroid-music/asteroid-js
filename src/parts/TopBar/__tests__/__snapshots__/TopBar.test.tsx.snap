// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> opens the correct sidebar when a tab button is pressed 1`] = `
<Material UI AppBar
  position="static"
>
  <div>
    Material UI Tabs with props:
    Object {
  "children": Array [
    &lt;Tab
      label="TestTab"
      value="TestTab"
    /&gt;,
  ],
  "onChange": [Function],
  "scrollButtons": "on",
  "value": false,
  "variant": "scrollable",
}
    <button
      onClick={[Function]}
    >
      Material UI Tab with props:
      Object {
  "label": "TestTab",
  "onChange": [Function],
  "value": "TestTab",
}
    </button>
  </div>
  <AsteroidSideBar
    closeCallback={[Function]}
    linkNames={
      Array [
        "TestSubtab",
        "TestSubtabTwo",
        "TestSubtab3",
        "TestSubtab-4",
      ]
    }
    open={true}
    tabName="TestTab"
  />
</Material UI AppBar>
`;

exports[`<TopBar /> renders correctly 1`] = `
<Material UI AppBar
  position="static"
>
  <div>
    Material UI Tabs with props:
    Object {
  "children": Array [
    &lt;Tab
      label="TestTab"
      value="TestTab"
    /&gt;,
  ],
  "onChange": [Function],
  "scrollButtons": "on",
  "value": false,
  "variant": "scrollable",
}
    <button
      onClick={[Function]}
    >
      Material UI Tab with props:
      Object {
  "label": "TestTab",
  "onChange": [Function],
  "value": "TestTab",
}
    </button>
  </div>
  <AsteroidSideBar
    closeCallback={[Function]}
    linkNames={Array []}
    open={false}
    tabName=""
  />
</Material UI AppBar>
`;
